// @nestjs/sequelize/dist/sequelize-core.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var SequelizeCoreModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const rxjs_1 = require("rxjs");
const sequelize_typescript_1 = require("sequelize-typescript");
const sequelize_utils_1 = require("./common/sequelize.utils");
const entities_metadata_storage_1 = require("./entities-metadata.storage");
const sequelize_constants_1 = require("./sequelize.constants");
let SequelizeCoreModule = SequelizeCoreModule_1 = class SequelizeCoreModule {
    constructor(options, moduleRef) {
        this.options = options;
        this.moduleRef = moduleRef;
    }
    static forRoot(options = {}) {
        const sequelizeModuleOptions = {
            provide: sequelize_constants_1.SEQUELIZE_MODULE_OPTIONS,
            useValue: options,
        };
        const connectionProvider = {
            provide: sequelize_utils_1.getConnectionToken(options),
            useFactory: () => __awaiter(this, void 0, void 0, function* () { return yield this.createConnectionFactory(options); }),
        };
        return {
            module: SequelizeCoreModule_1,
            providers: [connectionProvider, sequelizeModuleOptions],
            exports: [connectionProvider],
        };
    }
    static forRootAsync(options) {
        const connectionProvider = {
            provide: sequelize_utils_1.getConnectionToken(options),
            useFactory: (sequelizeOptions) => __awaiter(this, void 0, void 0, function* () {
                if (options.name) {
                    return yield this.createConnectionFactory(Object.assign(Object.assign({}, sequelizeOptions), { name: options.name }));
                }
                return yield this.createConnectionFactory(sequelizeOptions);
            }),
            inject: [sequelize_constants_1.SEQUELIZE_MODULE_OPTIONS],
        };
        const asyncProviders = this.createAsyncProviders(options);
        return {
            module: SequelizeCoreModule_1,
            imports: options.imports,
            providers: [
                ...asyncProviders,
                connectionProvider,
                {
                    provide: sequelize_constants_1.SEQUELIZE_MODULE_ID,
                    useValue: sequelize_utils_1.generateString(),
                },
            ],
            exports: [connectionProvider],
        };
    }
    onApplicationShutdown() {
        return __awaiter(this, void 0, void 0, function* () {
            const connection = this.moduleRef.get(sequelize_utils_1.getConnectionToken(this.options));
            connection && (yield connection.close());
        });
    }
    static createAsyncProviders(options) {
        if (options.useExisting || options.useFactory) {
            return [this.createAsyncOptionsProvider(options)];
        }
        const useClass = options.useClass;
        return [
            this.createAsyncOptionsProvider(options),
            {
                provide: useClass,
                useClass,
            },
        ];
    }
    static createAsyncOptionsProvider(options) {
        if (options.useFactory) {
            return {
                provide: sequelize_constants_1.SEQUELIZE_MODULE_OPTIONS,
                useFactory: options.useFactory,
                inject: options.inject || [],
            };
        }
        const inject = [
            (options.useClass || options.useExisting),
        ];
        return {
            provide: sequelize_constants_1.SEQUELIZE_MODULE_OPTIONS,
            useFactory: (optionsFactory) => __awaiter(this, void 0, void 0, function* () { return yield optionsFactory.createSequelizeOptions(options.name); }),
            inject,
        };
    }
    static createConnectionFactory(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield rxjs_1.defer(() => __awaiter(this, void 0, void 0, function* () {
                if (!options.autoLoadModels) {
                    return new sequelize_typescript_1.Sequelize(options);
                }
                const connectionToken = options.name || sequelize_constants_1.DEFAULT_CONNECTION_NAME;
                const sequelize = new sequelize_typescript_1.Sequelize(options);
                const models = entities_metadata_storage_1.EntitiesMetadataStorage.getEntitiesByConnection(connectionToken);
                sequelize.addModels(models);
                yield sequelize.authenticate();
                if (typeof options.synchronize === 'undefined' || options.synchronize) {
                    yield sequelize.sync(options.sync);
                }
                return sequelize;
            }))
                .pipe(sequelize_utils_1.handleRetry(options.retryAttempts, options.retryDelay))
                .toPromise();
        });
    }
};
SequelizeCoreModule = SequelizeCoreModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({}),
    __param(0, common_1.Inject(sequelize_constants_1.SEQUELIZE_MODULE_OPTIONS)),
    __metadata("design:paramtypes", [Object, core_1.ModuleRef])
], SequelizeCoreModule);
exports.SequelizeCoreModule = SequelizeCoreModule;
